module/zfs/vdev_trim.c

```diff
@@ -0,0 +1,1277 @@
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2016 by Delphix. All rights reserved.
+ * Copyright (c) 2019 by Lawrence Livermore National Security, LLC.
+ */
+
+#include <sys/spa.h>
+#include <sys/spa_impl.h>
+#include <sys/txg.h>
+#include <sys/vdev_impl.h>
+#include <sys/vdev_trim.h>
+#include <sys/refcount.h>
+#include <sys/metaslab_impl.h>
+#include <sys/dsl_synctask.h>
+#include <sys/zap.h>
+#include <sys/dmu_tx.h>
+
+/*
+ * Maximum size of TRIM command, ranges will be chunked in to 128MiB extents.
+ */
+unsigned int zfs_trim_extent_bytes_max = 128 * 1024 * 1024;
+
+/*
+ * Minimum size of TRIM commands, extents smaller than 32Kib will be skipped.
+ */
+unsigned int zfs_trim_extent_bytes_min = 32 * 1024;
+
+/*
+ * Maximum number of queued TRIMs outstanding per leaf vdev.  The number of
+ * concurrent TRIM commands issued to the device is controlled by the
+ * zfs_vdev_trim_min_active and zfs_vdev_trim_max_active module options.
+ */
+unsigned int zfs_trim_queue_limit = 10;
+
+/*
+ * How many transaction groups worth of updates should be aggregated before
+ * TRIM operations are issued to the device.  This setting represents a
+ * trade-off between issuing more efficient TRIM operations, by allowing
+ * them to be aggregated longer, and issuing them promptly enough that the
+ * space is trimmed and available for use by the device.
+ *
+ * Increasing this value will allow frees to be aggregated for a longer
+ * time.  This will result is larger TRIM operations, and increased memory
+ * usage in order to track the pending TRIMs.  Decreasing this value will
+ * have the opposite effect.  The default value of 32 was determined to be
+ * a reasonable compromise.
+ */
+int zfs_trim_txg_batch = 32;
+
+typedef struct trim_args {
+	vdev_t		*trim_vdev;
+	metaslab_t	*trim_msp;
+	range_tree_t	*trim_tree;
+	trim_type_t	trim_type;
+	hrtime_t	trim_start_time;
+	uint64_t	trim_bytes_done;
+	uint64_t	trim_extent_bytes_max;
+	uint64_t	trim_extent_bytes_min;
+	enum trim_flag	trim_flags;
+} trim_args_t;
+
+static boolean_t
+vdev_trim_should_stop(vdev_t *vd)
+{
+	return (vd->vdev_trim_exit_wanted || !vdev_writeable(vd) ||
+	    vd->vdev_detached || vd->vdev_top->vdev_removing);
+}
+
+/*
+ * Determines the minimum sensible rate at which a manual TRIM can be
+ * performed on a given spa and returns it (in bytes per second).  The
+ * minimum rate is  calculated by assuming that trimming a metaslab
+ * should not take longer than 1000 seconds.
```

Comment:
That seems pretty arbitrary.  Why do we need a lower bound at all?  Given that the number of metaslabs could vary considerably, should this be based on time to trim a vdev instead?

Reply:
I'm not convinced really we need a lower bound at all, this was a hold over from the previous PR.  I've replaced this code for now with a minimum 1MB/s which can be requested, I'd be OK with removing this minimum entirely.

Reply:
I would also be more in favor of removing the lower bound.